src/expr/impl/src/scalar/jsonb_path.rs

```diff
@@ -0,0 +1,324 @@
+)]
+fn jsonb_path_exists4(
+    target: JsonbRef<'_>,
+    vars: JsonbRef<'_>,
+    silent: bool,
+    path: &JsonPath,
+) -> Result<Option<bool>> {
+    match jsonb_path_exists3(target, vars, path) {
+        Ok(x) => Ok(Some(x)),
+        Err(_) if silent => Ok(None),
+        Err(e) => Err(e),
+    }
+}
+
+#[function(
+    "jsonb_path_match(jsonb, varchar) -> boolean",
+    prebuild = "JsonPath::new($1).map_err(parse_error)?"
+)]
+fn jsonb_path_match2(target: JsonbRef<'_>, path: &JsonPath) -> Result<Option<bool>> {
+    let matched = path
+        .query::<ValueRef<'_>>(target.into())
+        .map_err(eval_error)?;
+
+    if matched.len() != 1 || !matched[0].as_ref().is_boolean() && !matched[0].as_ref().is_null() {
+        return Err(ExprError::InvalidParam {
+            name: "jsonb_path_match",
+            reason: "single boolean result is expected".into(),
+        });
+    }
+    Ok(matched[0].as_ref().as_bool())
+}
+
+#[function(
+    "jsonb_path_match(jsonb, varchar, jsonb) -> boolean",
+    prebuild = "JsonPath::new($1).map_err(parse_error)?"
+)]
+fn jsonb_path_match3(
+    target: JsonbRef<'_>,
+    vars: JsonbRef<'_>,
+    path: &JsonPath,
+) -> Result<Option<bool>> {
+    let matched = path
+        .query_with_vars::<ValueRef<'_>>(target.into(), vars.into())
+        .map_err(eval_error)?;
+
+    if matched.len() != 1 || !matched[0].as_ref().is_boolean() && !matched[0].as_ref().is_null() {
+        return Err(ExprError::InvalidParam {
+            name: "jsonb_path_match",
+            reason: "single boolean result is expected".into(),
+        });
+    }
+    Ok(matched[0].as_ref().as_bool())
+}
+
+/// Returns the result of a JSON path predicate check for the specified JSON value.
+/// Only the first item of the result is taken into account.
+/// If the result is not Boolean, then NULL is returned.
+/// The optional vars and silent arguments act the same as for `jsonb_path_exists`.
+///
+/// # Examples
+///
+/// ```slt
+/// query B
+/// select jsonb_path_match('{"a":[1,2,3,4,5]}', 'exists($.a[*] ? (@ >= $min && @ <= $max))', '{"min":2, "max":4}');
+/// ----
+/// t
+/// ```
+#[function(
+    "jsonb_path_match(jsonb, varchar, jsonb, boolean) -> boolean",
+    prebuild = "JsonPath::new($1).map_err(parse_error)?"
+)]
+fn jsonb_path_match4(
+    target: JsonbRef<'_>,
+    vars: JsonbRef<'_>,
+    silent: bool,
+    path: &JsonPath,
+) -> Result<Option<bool>> {
+    match jsonb_path_match3(target, vars, path) {
+        Ok(x) => Ok(x),
+        Err(_) if silent => Ok(None),
+        Err(e) => Err(e),
+    }
+}
+
+#[function(
+    "jsonb_path_query(jsonb, varchar) -> setof jsonb",
+    prebuild = "JsonPath::new($1).map_err(parse_error)?"
+)]
+fn jsonb_path_query2<'a>(
+    target: JsonbRef<'a>,
+    path: &JsonPath,
+) -> Result<impl Iterator<Item = JsonbVal> + 'a> {
+    let matched = path
+        .query::<ValueRef<'_>>(target.into())
+        .map_err(eval_error)?;
+    Ok(matched.into_iter().map(|json| json.into_owned().into()))
```

Comment:
Is it possible to return an iterator for `query`?

Reply:
It is hard in stable Rust as all query functions need to be written as generator.

Reply:
What about those 3rd party libraries available under `stable` 🥵? Like https://github.com/Wolvereness/remit/

Reply:
Interesting! Will have a try later.